lib/vrl/compiler/src/type_def.rs

```diff
@@ -176,7 +176,12 @@ impl KindInfo {
                     self = KindInfo::Known(set);
                 }
                 Segment::Index(index) => {
-                    let index = *index as usize;
+                    let index = if index.is_negative() {
+                        0
+                    } else {
+                        *index as usize
+                    };
+
```

Comment:
I'm not sure this is right, but it could just be my misunderstanding.

If you set a negative index: `.foo[-1] = 1` is this code setting the type at index 0 for that path? Surely that's the wrong index to set and if it can't work out the correct one, it should not set any at all or wipe all types for that array?

Reply:
Yeah, that's correct, we don't know the number of items in the array, so there's no way to know what to do with a negative index. I chose instead to treat `-1` as being the first in the array.

However, I agree It's a trade-off, and I also agree with you that the current trade-off isn't the best here.

I _think_ we can do better here, by using `self.at_path(path)` to see if we have any information about the array, and then try and update that information.

However, I'll probably start by setting the type def to "unknown" for now, and then iterate on this in the future.